# Optical character recognition (OCR) 
Project 4 Group 3


## Error Detection

First of all, we need to detect errors, or incorrectly processed words. Here we extract features according to the paper and use SVM for garbage detection.

In [1]:
import os
import glob
import re
from collections import Counter
import pandas as pd
import collections
import numpy as np
from sklearn.svm import SVC
import itertools
#import itertools
#nltk.download('punkt')

In [2]:
##### read ground_truth
ground_dir = glob.glob(os.path.join(os.getcwd(),'../data/ground_truth','*.txt'))
ground_tokens = []
for gd in ground_dir:
    with open(gd) as ground_file:
        ground_raw = ground_file.read()
        ground_t = ground_raw.split()   
        ground_tokens += ground_t
print(len(ground_tokens))
print(ground_tokens[:20])

294841
['communications', 'network.', 'Member', 'companies', 'are', 'strongly', 'encouraged', 'to', 'provide', 'this', 'needed', 'support.', 'The', 'state', 'advocacy', 'program*', 'including', 'the', 'new', 'CMA/LINC']


In [3]:
tess_dir = glob.glob(os.path.join(os.getcwd(),'../data/tesseract','*.txt'))
#ground_dir = glob.glob(os.path.join(os.getcwd(),'data/ground_truth','*.txt'))
file_name_gd = []
file_name_td = []
for gd, td in zip(ground_dir, tess_dir):
        with open(gd, encoding="utf8") as ground_file:    #, encoding="utf8"
            with open(td, encoding="utf8") as tess_file:                
                ground_r = list(ground_file.readlines()) 
                tess_r = list(tess_file.readlines())
                if len(tess_r) == len(ground_r):
                    file_name_td.append(td)
                    file_name_gd.append(gd)

In [4]:
# make sure all the lines have the same size.
ground_tokens=[]
tess_tokens=[]
for gd, td in zip(file_name_gd, file_name_td):
        with open(gd) as file1:
            with open(td) as file2:
                for line1,line2 in zip(file1,file2):
                    if len(line1)==len(line2):
                        for word1,word2 in zip(line1.split(),line2.split()):
                            ground_tokens.append(word1)
                            tess_tokens.append(word2)

In [5]:
print(tess_tokens[:30])
print(ground_tokens[:30])

['communlcatlons', 'network.', 'Member', 'companles', 'are', 'strongly', 'encouraged', 'to', 'provlde', 'thls', 'needed', 'support.', 'The', 'state', 'advocacy', 'program"', '1nclud1ng', 'the', 'new', 'CMA/LINC', 'computer', 'network.', 'will', 'be', 'heavlly', '1nvolved', '1n', '1995', '1n', 'the']
['communications', 'network.', 'Member', 'companies', 'are', 'strongly', 'encouraged', 'to', 'provide', 'this', 'needed', 'support.', 'The', 'state', 'advocacy', 'program*', 'including', 'the', 'new', 'CMA/LINC', 'computer', 'network,', 'will', 'be', 'heavily', 'involved', 'in', '1986', 'in', 'the']


In [6]:
print(len(tess_tokens))
print(len(ground_tokens))

152143
152143


In [7]:
y = []
for gt, tt in zip(ground_tokens, tess_tokens):
        if gt == tt:
            y.append(0)   # 0 indicates correct
        else:
            y.append(1)   # 1 indicates error

In [8]:
X = pd.DataFrame(tess_tokens)   
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20)
ground_train = [ground_tokens[i] for i in X_train.index.tolist()]
ground_test = [ground_tokens[i] for i in X_test.index.tolist()]
X_train = X_train[0].tolist()
X_test = X_test[0].tolist()

In [9]:
print(X_train[:30])
print(ground_train[:30])
print(y_train[:30])

['esent', 'of', 'for', 'tax', 'Superfund', 'present', 'changes', 'exposure', '11m:', 'Congress', '£111', '1n', '1991792', 'hazardous', 'the', '15', 'of', '15', 'members.', "1ndustry's", '35515:', 'protectlon', '1dent1ty', 'progress', 'an', 'The', 'classes', 'he', 'on', 'Vlew']
['non-degredation', 'of', 'for', 'tax', 'Superfund', 'present', 'changes', 'exposure', 'zinc', 'Congress', 'fill', 'in', '1981-82', 'hazardous', 'the', 'is', 'of', 'is', 'members.', "industry's", 'assist', 'protection', 'identity', 'progress', 'an', 'The', 'classes', 'he', 'on', 'view']
[1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1]


In [10]:
# define character type
vowels = list('aeiou')
consonants = list("bcdfghjklmnpqrstvwxyz")
digits = list('0123456789')

def count_up(word):
    u = [x for x in word if x.isupper()]
    return len(u)

def count_low(word):
    l = [x for x in word if x.islower()]
    return len(l)

# count the number of consecutive of the same symbol (if greater than 3 => garbadge)
def count_cons_occur(word):
    max_occ=0
    previous=None
    temp=1
    for i in word:
        if i==previous:
            temp=temp+1
        else:
            max_occ=max(max_occ,temp)
            temp=1
        previous=i
    return(max_occ)

#  count the number of consecutive of the consonants(if greater than 6 => garbadge)
def count_cons_occur_consonants(word):
    max_occ=0
    temp=0
    for i in word:
        if i in consonants:
            temp=temp+1
        else:
            max_occ=max(max_occ,temp)
            temp=0
    return(max_occ)

# get the subset of the word: delete the 1st and last word 
def subset_word(word):
    return word[1:len(word) - 1]

#  do the division in case of  prevent from divisor = 0        
def div(x,y):
    if y != 0:
        return x/y
    return 0

In [11]:
# get the bigram frequency for the particular word (return the array of each bf_i)
def bigram_freq(word, bi_dict):
    word = word.lower()
    bf = []
    for i in range(len(word)-1):
        key = word[i:i+2]
        bf.append(bi_dict[key])
    return(bf)

# Form the list of letter bigrams from our ground truth dataset -- bi_dict
# First we normalized all tokens to the lowercase 
lower_case_words = []
for word in ground_tokens:
    lower_word = word.lower()
    lower_case_words.append(lower_word)

# Secondly, we form the LB from our ground_truth dataset 
# create a dictionary-like collection: key: bigrams value: frequency 
LB = collections.defaultdict(int)
for word in lower_case_words:
    for i in range(len(word)-1):
        key = word[i:i+2]
        LB[key] +=1

In [16]:
def feature_matrix(data):
    ##### Feature extracting
    
    # feature 1 : length of the each words
    length = []

    # feature 2: number of each vowels and consonants; and the corresponding ratio
    vowels_count = []
    consonants_count = []
    ratio_v_l = []
    ratio_c_l = []
    ratio_v_c = []

    # feature 3: number of special symbols: not vowels, not consonants and not digits 
    nonalpha = []

    # feature 4: number of digits and the ratio bewteen number of digit and length of the words 
    digits = []
    ratio_d_l = []

    # feature 5: number of lowercase chars, uppercase chars and corresponding ratio 
    lower_case = []
    upper_case = []
    ratio_low_l = []
    ratio_up_l = []
    
    # feature 6: number of consecutive occurances of the same symbol and the ratio between it and the length
    consecutive_occur = []
    ratio_sym_l = []
    
    # feature 7: number of all alpha(words) and numbers in the words and the ratio between it and the length
    # total number of vowels, consonants and digits
    ratio_la = [] 
    
    # feature 8: number of consecutive consonants in the word
    consonants_consecutive = []
    
    # feature 9: delete the 1st and the last symbol and then check the number of non alpha - number  character 
    infix_nonalpha = []

    # feature 10: bigram frequency 
    bigram = []
    
    # feature 11: number of most frequent symbols (including alpha,numbers, special symbols)
    most_freq = []
    
    # feature 12: number of non-alpha symbols: length(words) - length(alpha)
    l2 = []
    
    
        # feature 1:
    for word in data:
        L= len(word)
        length.append(L)

        # feature 2: vowels
        word_lower = word.lower()
        v_count = len([v for v in word if v in vowels])
        vowels_count.append(v_count)
        
        # feature 2: consonants
        c_count = len([c for c in word if c in consonants])
        consonants_count.append(c_count)
   
        ratio_v_l.append(v_count/L)
        ratio_c_l.append(c_count/L)
        ratio_vc = div(v_count,c_count)
        ratio_v_c.append(ratio_vc)

        # feature 3
        symbol_count = len([s for s in word if s not in vowels or consonants or digits])
        nonalpha.append(symbol_count)
     
        # feature 4
        d_count = len([d for d in word if d in digits])
        digits.append(d_count)
        ratio_d_l.append(d_count/L)
        
        # feature 5
        low_count = count_low(word)
        lower_case.append(low_count)
        upper_count = count_up(word)
        upper_case.append(upper_count)

        ratio_low_l.append(low_count/L)
        ratio_up_l.append(upper_count/L)

        # feature 6
        cons_occur_count = int(count_cons_occur(word))
        consecutive_occur.append(cons_occur_count)
        if cons_occur_count >= 3:
            ratio_sym_l.append(cons_occur_count/L)
        else:
            ratio_sym_l.append(0)

        # feature 7
        la = L-symbol_count
        if symbol_count > la:
            ratio_la.append(1)
        else:
            ratio_la.append(0)

        # feature 8
        consonance_consecu_count = int(count_cons_occur_consonants(word))
        if consonance_consecu_count >= 6:
            consonants_consecutive.append(1)
        else:
            consonants_consecutive.append(0)

        # feature 9 
        # delete the 1st and last symbol of the string: remaining(2 or more special symbol) = 1; otherwise, = 0
        word_removed = subset_word(word)
        remove_count = len([r for r in word_removed if r not in vowels or consonants or digits])
        if remove_count >=3:
            infix_nonalpha.append(1)
        else:
            infix_nonalpha.append(0)  

        # feature 10 bigram
        bf = bigram_freq(word, LB)
    
        lower_tess_tokens = []
        #  = word.lower()
        lower_tess_tokens.append(word_lower)
        
        # number of bigrams of the string 
        n = len(lower_tess_tokens)-1
        # bigr: (sum(bf_i)/10000) / n
        big = div((sum(bf)/10000),n)
        bigram.append(big)

        # feature 11 most frequent symbol
        i_count = Counter(word).most_common(1)[0][1]
        if i_count >=3:
            most_freq.append(i_count/L)   # i/l
        else:
            most_freq.append(0)

        # feature 12 Non-alphabetical symbols: nonalpha/total symbols
        l1 = len([v for v in word_lower if v in vowels]) + len([c for c in word_lower if c in consonants])
        l2_count = L - l1
        ratio_l2 = div(l2_count,l1)
        l2.append(ratio_l2)  
        
    ##### construct a feature dataframe for SVM
    df1 = pd.DataFrame({'length': length,
                        'vowels': vowels_count,
                        'consonants': consonants_count,
                        'quot v/l': ratio_v_l,
                        'quot c/l': ratio_c_l,
                        'quot v/c': ratio_v_c,
                        'nonalpha': nonalpha,
                        'digits': digits,
                        'quot d/l': ratio_d_l,
                        'lowers': lower_case,
                        'uppers': upper_case,
                        'quot low/l': ratio_low_l,
                        'quot up/l': ratio_up_l,
                        'cons occur': consecutive_occur,
                        'quot cons/l': ratio_sym_l,
                        'quot la': ratio_la,
                        'consonants_secut': consonants_consecutive,
                        'fix_nonalpha': infix_nonalpha,
                        'bigr': bigram,
                        'most_freq' : most_freq,
                        'l2': l2})
    return df1

In [20]:
X_feature_train = feature_matrix(X_train)
X_feature_test = feature_matrix(X_test)

In [21]:
svclassifier = SVC(kernel='rbf')  
svclassifier.fit(X_feature_train, y_train)  

/Users/shengweihuang/anaconda3/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [22]:
import pickle
from google.colab import files

filename = 'svm_model.pkl'
pickle.dump(svclassifier, open(filename, 'wb'))
files.download('svm_model.pkl')
# Can load the model from output or upload to colab
# svm_model = pickle.load(open(filename, 'rb'))

ModuleNotFoundError: No module named 'google'

In [ ]:
#to predict
y_pred = svclassifier.predict(X_feature_test)  
# y_pred = svm_model.predict(X_feature_test)  

In [ ]:
df_output = pd.DataFrame({'tokens_tesseract':X_test,
                          'Predict_by_SVM': y_pred})
print(df_output[:100])

In [ ]:
##### evaluation
#confustion Matrix
from sklearn.metrics import classification_report, confusion_matrix  
print(confusion_matrix(y_test,y_pred))  
print(classification_report(y_test,y_pred))  

## Error correction

Given the detected word error, in order to find the best correction, we need to generating the candidate corrections: a dictionary or a database of legal n-grams to locate one or more potential correction terms. Then we need invoke some lexical-similarity measure between the misspelled string and the candidates or a probabilistic estimate of the likelihood of the correction to rank order the candidates. Here, we implement the positional binary digram method in the first reference paper. (https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=1672564}{positional binary digram)

In [0]:
import string

In [0]:
def keep_alphabet(tokens):
  # only retain alphabet
  out = []
  for l in tokens:
      l = l.lower()
      if l in set('abcdefghijklmnopqrstuvwxyz '):
          out.append(l)
  return ''.join(out)

In [0]:
# Replace a postion of a string
def replace_str_index(text,index=0,replacement=''):
    return '%s%s%s'%(text[:index],replacement,text[index+1:])

In [0]:
# Define the main correction function
def correction(word, digrams):
  detect = 0
  beta = []
  # Find the error positions
  for each in digrams:
    matrix = digrams[each]
    if matrix[string.ascii_lowercase.index(word[each[0]])][string.ascii_lowercase.index(word[each[1]])] == 0:
      detect += 1
      beta_each = set(each)
      if detect == 1:
        beta = beta_each
      else:
        beta = beta.intersection(beta_each)
  # print(beta)
  # Consider when there is one or two elements in beta
  if len(beta) in [1,2]:
    choices = 0
    v_dict = {}
    for i in beta:
      v_list = []
      position = i
      for j in range(len(word)):
        alpha_j = string.ascii_lowercase.index(word[j])
        if j < position:
          vector_j = digrams[(j, position)][alpha_j]
          v_list.append(vector_j)
        elif j > position:
          vector_j = [item[alpha_j] for item in digrams[(position, j)]]
          v_list.append(vector_j)
      v = v_list[0]
      for each in v_list:
        v = [a and b for a, b in zip(v, each)]
      if sum(v) == 1:
        choices += 1
        v_dict[i] = v
    if choices == 1:
      for key in v_dict:
        position_chosen = key
      word = replace_str_index(word, position_chosen, string.ascii_lowercase[v_dict[position_chosen].index(1)])
  return word

In [0]:
for i in range(len(ground_train)):
  ground_train[i] = keep_alphabet(ground_train[i])

In [0]:
# Create dictionary of digrams
token_by_len = collections.defaultdict(list)
digrams_by_len = collections.defaultdict(dict)
for w in ground_train:
  token_by_len[len(w)].append(w)
  
#print('Number of words of diffenrent length:')
#for key, value in token_by_len.items() :
#    print (key, len(value))

for length in token_by_len:
  for i, j in itertools.combinations(range(length), 2):
    key = (i, j)
    matrix = [[0] * 26 for _ in range(26)]
    for words in token_by_len[length]:
      matrix[string.ascii_lowercase.index(words[i])][string.ascii_lowercase.index(words[j])] = 1
    digrams_by_len[length][key] = matrix

In [90]:
# Number of 1 in digrams with given length
ae = 0
for each in digrams_by_len[10][(0,1)]:
  ae += sum(each)
print(ae)

144


In [0]:
# Clean the words
corrected_test = X_test.copy()
for i in range(len(corrected_test)):
  corrected_test[i] = keep_alphabet(corrected_test[i])
for i in range(len(ground_test)):
  ground_test[i] = keep_alphabet(ground_test[i])

In [0]:
# Make correction
for i in range(len(y_pred)):
  if y_pred[i] == 1:
    word_length = len(corrected_test[i])
    if word_length > 1:
      digrams_i = digrams_by_len[word_length]
      corrected_test[i] = correction(corrected_test[i], digrams_i)

In [0]:
y_corrected = []
for gt, ct in zip(ground_test, corrected_test):
        if gt == ct:
            y_corrected.append(0)   # 0 indicates correct
        else:
            y_corrected.append(1)   # 1 indicates error

In [148]:
# Compare the results
print(corrected_test[:30])
print(ground_test[:30])
print(X_test[:30])

['n', 'on', 'and', 'washlngton', '', '', 'report', '', 'reached', 'could', 'cozporatlon', 'the', 'recent', 'vent', 'sesslon', 'are', 'and', 'able', 'between', 'commodlty', 'the', 'mr', 'addltlonal', 'to', 'approach', 'a', 'commlttee', 'are', 'would', 'but']
['in', 'on', 'and', 'washington', '', 'it', 'report', '', 'reached', 'could', 'corporation', 'the', 'recent', 'vent', 'session', 'are', 'and', 'able', 'governments', 'commodity', 'the', 'mr', 'additional', 'to', 'approach', 'a', 'committee', 'are', 'would', 'but']
['1n', 'on', 'and', 'Washlngton,', '7', '1:', 'Report', '40', 'reached', 'could', 'Cozporatlon', 'the', 'recent', 'Vent', 'sesslon', 'are', 'and', 'able', 'between', 'commodlty', 'the', 'Mr.', 'Addltlonal', 'to', 'approach', 'a', 'Commlttee', 'are:', 'would', 'but']


In [149]:
print(sum(y_test))
print(sum(y_corrected))

12039
10362


## Performance measure

In [0]:
# Create corresponding lists of characters
corrected_char = []
for each in corrected_test:
  corrected_char += each
corrected_char

ground_char = []
for each in ground_test:
  ground_char += each

tess_char = [] 
tess_test = X_test.copy()
for i in range(len(tess_test)):
  tess_test[i] = keep_alphabet(tess_test[i])
for each in tess_test:
  tess_char += each

In [0]:
import numpy as np

In [0]:
# Count the number of characters
ground_count = []
corrected_count = []
tess_count = []
for each in string.ascii_lowercase:
  ground_count.append(ground_char.count(each))
  corrected_count.append(corrected_char.count(each))
  tess_count.append(tess_char.count(each))
ground_corrected_min = np.minimum(ground_count, corrected_count)
ground_tess_min = np.minimum(ground_count, tess_count)

In [188]:
# Make the evaluation table
word_recall_tess = 1 - sum(y_test)/len(ground_test)
word_recall_corrected = 1 - sum(y_corrected)/len(ground_test)
word_precision_tess = 1 - sum(y_test)/len(X_test)
word_precision_corrected = 1 - sum(y_corrected)/len(X_test)
char_recall_tess = sum(ground_tess_min)/sum(ground_count)
char_precision_tess = sum(ground_tess_min)/sum(tess_count)
char_recall_corrected = sum(ground_corrected_min)/sum(ground_count)
char_precision_corrected = sum(ground_corrected_min)/sum(corrected_count)

d = {'Tesseract': [word_recall_tess, word_precision_tess, char_recall_tess, char_precision_tess], 
     'Tesseract_with_postprocessing': [word_recall_corrected, word_precision_corrected 
                                       ,char_recall_corrected, char_precision_corrected]}
OCR_performance_table = pd.DataFrame(data=d)
OCR_performance_table.rename(index={0: 'word_wise_recall', 1:'word_wise_precision', 
                                    2:'character_wise_recall', 3:'character_wise_precision'}, inplace=True)
print(OCR_performance_table)

                          Tesseract  Tesseract_with_postprocessing
word_wise_recall           0.604358                       0.659470
word_wise_precision        0.604358                       0.659470
character_wise_recall      0.912326                       0.917584
character_wise_precision   0.945590                       0.951040


The word-wise recall after postprocessing is 0.6595, significantly better than original Tesseract text. Word-wise precision is the same as recall because of our preprocessing. The character-wise recall and precision are slightly improved to 0.9176 and 0.951.